[DATA-605] Orbslam homebrew formula

tessavitabile opened the pull request:
https://viam.atlassian.net/browse/DATA-605

A couple questions:
- After running `brew install`, I still had to manually link everything from `/opt/homebrew/Cellar/orb-grpc-server/HEAD-af2dec0/lib/` into `/usr/local/lib/`. Otherwise, I got a runtime error:
`dyld[13090]: Library not loaded: /tmp/orb-grpc-server-20221026-10229-f8sjnr/slam-libraries/viam-orb-slam3/ORB_SLAM3/Thirdparty/DBoW2/lib/libDBoW2.dylib
  Referenced from: /opt/homebrew/Cellar/orb-grpc-server/HEAD-af2dec0/bin/orb_grpc_server
  Reason: tried: '/tmp/orb-grpc-server-20221026-10229-f8sjnr/slam-libraries/viam-orb-slam3/ORB_SLAM3/Thirdparty/DBoW2/lib/libDBoW2.dylib' (no such file), '/usr/local/lib/libDBoW2.dylib' (no such file), '/usr/lib/libDBoW2.dylib' (no such file)`
I'm not sure if this is a problem with HEAD installs, or a problem with the brew formula that I wrote. Any advice?
**EDIT**: Now that I recall, I had the same problem when I installed `pangolin` using the brew package.
- I tested on my M1. Do we have access to an intel mac to test on?
**UPDATE**: Kat and I tested on her intel mac, and the behavior was mostly the same as on my M1 mac. Yay! The only difference was that on my mac, the binary was installed to `/opt/homebrew/bin`, so `which orb_grpc_server` returned something. On Kat's mac, the binary wasn't linked to any location that we found, so `which orb_grpc_server` returned nothing. We tested by running the binary in the Cellar directly.

tessavitabile commented on Formula/orb-grpc-server.rb:
```diff
@@ -0,0 +1,32 @@
+class OrbGrpcServer < Formula
+  desc "A Viam slam GRPC server for ORB_SLAM3"
+  homepage "https://www.viam.com/"
+  url "https://github.com/viamrobotics/slam/archive/refs/tags/v0.1.4.tar.gz"
```
Do you know if we have a plan for updating these urls?

Otterverse replied:
We don't. Someone has to update this formula when it's time to update. There's a plan to automate that for RDK (viam-server) but it's not done yet.

Otterverse reviewed (changes requested):
So I tested this on my linux machine and it fails to build, because the setup script tries to call brew from inside brew (during one of the make install steps) and that's not allowed. It should never do this, as it should simply depend on other brew formula directly for dependencies.

Additionally, I tested on mac, and it also failed there because it tries to do a sudo operation.

A brew formula needs to install EVERYTHING it needs using only the formula itself, or dependencies of other brew formulae. A formula should never be installing system level packages directly with apt, or elevating privileges with sudo or similar.

tessavitabile commented:
Sorry, I should have said, this requires the changes from https://github.com/viamrobotics/slam/pull/95. Can you try change `branch` to `main` (since the PR is now merged, and I deleted the branch) and installing with `--HEAD`?

tessavitabile commented:
Now that https://github.com/viamrobotics/slam/pull/95 is merged, I changed `branch` to `main`. For now, this will only work with `--HEAD`. Once I tag a version that includes https://github.com/viamrobotics/slam/pull/95, I'll update `url` to point to that tagged version.

Otterverse commented:
I've been looking into the issue you call out in top comment/body of this PR, about the libraries not being in the correct locations.

During the build this comes up:
```CMake Warning (dev):
  Policy CMP0042 is not set: MACOSX_RPATH is enabled by default.  Run "cmake
  --help-policy CMP0042" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  MACOSX_RPATH is not specified for the following targets:

   ORB_SLAM3
```
The RPATH stuff is key here I think. The "proper" way to do this with cmake is to actually setup install targets and handle the "install" from within cmake itself, with rpath stuff enabled. Cmake does a lot of "magic" during the install step to swap paths around in built binaries. Then, in brew, call cmake directly with *std_cmake_args, which will handle setting up all the paths and details for the build in brew, and use cmake's install rather than directly picking out the libraries and main executable. There's an entire article on the cmake wiki about dealing with RPATH issues if you wanted to go that route. https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/RPATH-handling

That's a lot of work though, even though it'd be the "correct" way of dealing with things. With that in mind, I found a simpler (but uglier) approach that still works. Basically, just swaping out the broken build-time library paths with their post-install paths. OSX has a tool for this called "install_name_tool." Adding the following to the formula (after the build but before the install steps) appears to fix it for me.

```
      if OS.mac?
        system "install_name_tool", "-change", buildpath.to_s.delete_prefix("/private") + "/slam-libraries/viam-orb-slam3/ORB_SLAM3/Thirdparty/DBoW2/lib/libDBoW2.dylib", "#{lib}/libDBoW2.dylib", "viam-orb-slam3/ORB_SLAM3/lib/libORB_SLAM3.dylib"
        system "install_name_tool", "-change", buildpath.to_s.delete_prefix("/private") + "/slam-libraries/viam-orb-slam3/ORB_SLAM3/Thirdparty/g2o/lib/libg2o.dylib", "#{lib}/libg2o.dylib", "viam-orb-slam3/ORB_SLAM3/lib/libORB_SLAM3.dylib"      
        system "install_name_tool", "-change", buildpath.to_s.delete_prefix("/private") + "/slam-libraries/viam-orb-slam3/ORB_SLAM3/Thirdparty/DBoW2/lib/libDBoW2.dylib", "#{lib}/libDBoW2.dylib", "viam-orb-slam3/bin/orb_grpc_server"
        system "install_name_tool", "-change", buildpath.to_s.delete_prefix("/private") + "/slam-libraries/viam-orb-slam3/ORB_SLAM3/Thirdparty/g2o/lib/libg2o.dylib", "#{lib}/libg2o.dylib", "viam-orb-slam3/bin/orb_grpc_server"
        system "install_name_tool", "-change", buildpath.to_s.delete_prefix("/private") + "/slam-libraries/viam-orb-slam3/ORB_SLAM3/lib/libORB_SLAM3.dylib", "#{lib}/libORB_SLAM3.dylib", "viam-orb-slam3/bin/orb_grpc_server"
      end
```

I tested on linux and an M1 mac. Would be good to test on an intel mac as well, make sure it works there too.

tessavitabile commented:
Thank you, @Otterverse, both for the longer explanation and the easier solution. I had started reading about RPATH when trying to fix openssl for the macOS CI job, and that does seem like it would take me awhile to get right, so I'm using the solution you gave.

I also updated the tag (and switched the download to a git clone, so that submodules are included), so it's now possible to test installing without `--HEAD`.

@kkufieta, would you be willing to test the formula again with these changes? You should now be able to install it with `brew install --build-from-source --verbose --debug orb-grpc-server`. I'm interested in whether we see the `Library not loaded` error when we try to run `orb_grpc_server`.

Otterverse reviewed (approved):
Woot! Works cleanly on my m1 Macbook and amd64 Linux desktop. If it clears the intel mac test, good to go I think!

kkufieta reviewed (approved):
Tested on intel mac, it works! :)
